client/comms/wsconn.go

```diff
@@ -165,18 +174,18 @@ func NewWsConn(cfg *WsCfg) (WsConn, error) {
 func (conn *wsConn) IsDown() bool {
 	conn.connectedMtx.RLock()
 	defer conn.connectedMtx.RUnlock()
-	return !conn.connected
+	return conn.connectionStatus != Connected
 }
 
-// setConnected updates the connection's connected state and runs the
+// setConnectionStatus updates the connection's status and runs the
 // ConnectEventFunc in case of a change.
-func (conn *wsConn) setConnected(connected bool) {
+func (conn *wsConn) setConnectionStatus(status ConnectionStatus) {
 	conn.connectedMtx.Lock()
-	statusChange := conn.connected != connected
-	conn.connected = connected
+	statusChange := conn.connectionStatus != status
+	conn.connectionStatus = status
```

Comment:
It looks like you could change this to an atomic an remove the lock if you wanted.

Reply:
Obviously it's very unlikely, but wouldn't it be possible to call `ConnectEventFunc` twice for the same status update, if two goroutines are updating to the same status it the same time?